pytrip/concave_tool.py

```diff
@@ -0,0 +1,302 @@
+import math
+
+# This file contains methods that create from a list of intersections (list of points).
+# Algorithm steps:
+#     1. Divide passed list of intersections into separate groups, that each represents a valid contour.
+#     2. For each group create a contour.
+#         2.1. Divide group of points into parts of contour, each part imitates discrete function (like in maths)
+#         2.2. Connect parts of contour (taking into account direction of connection)
+#     3. Return created contours.
+# Basic heuristics are used:
+#     1. searching for minimal distance between a part and a point and ensuring that relation is symmetrical
+#         (if point A is the closest one to part B, check if part B is the closest one to point A)
+#     2. limiting distance between parts and points based on previous distances
+#         (next point cannot be too far away from parts)
+
+
+class ListEntry:
+    """Special type to hold data and other useful information"""
+    def __init__(self):
+        self.data = []
+        self.last_distance = float('inf')
+
+
+class SpecialPoint:
+    """Special type to hold data and other useful information"""
+    def __init__(self, point):
+        self.point = point
+        self.is_appended = False
+
+
+# -------------------- start block of utility methods --------------------
+def map_points_to_special_points(points):
+    mapped = []
+    for p in points:
+        mapped.append(SpecialPoint(p))
+    return mapped
+
+
+def calculate_distance(points_a, point_b):

```

Comment:
I'd rename arguments to `p` and `q` and add short docstring mentioning we calculate distance in Euclidean 3D space